lib/wit/package.py

```diff
@@ -1,73 +1,149 @@
+        self.name = name
+        self.source = source
+        self.unresolved_revision = unresolved_revision or "HEAD"
+        self.revision = None
+
+        self.repo = None
+
+        self.find_source(repo_paths)
+
+        self.dependents = []
+
+    def short_revision(self):
+        rev = self.revision or self.unresolved_revision
+        if len(rev) >= 40:
+            rev = rev[:8]
+        return rev
+
+    def __key(self):
+        return (self.source, self.unresolved_revision, self.name)
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __eq__(self, other):
+        return isinstance(self, type(other)) and self.__key() == other.__key()
+
+    def add_dependent(self, dep):
+        if dep not in self.dependents:
+            self.dependents.append(dep)
+
+    def load_repo(self, wsroot, download=False):
+        """Connect a Package to a GitRepo on disk.
+
+        If found, self.repo will be updated
+        and self.revision will be updated to the resolved self.unresolved_revision
         """
-        # TODO Could speed up validation
-        #   - use git ls-remote to validate remote exists
-        #   - use git ls-remote to validate revision for tags and branches
-        #   - if github repo, check if page exists (or if you get 404)
-        # FIXME: This is ugly. Split on '::' into a path and revision, but
-        # there may not be a revision. So add an additional array
-        source, rev = (s.split("::") + [None])[:2]
-        if rev is None:
-            rev = "HEAD"
-
-        return gitrepo.GitRepo(source, rev)
-
-    @staticmethod
-    def from_manifest(wsroot, m):
-        commit = m['commit']
-        name = m['name']
-
-        # Source is not required, because the repo may be found on
-        # $WIT_REPO_PATH
-        source = m.get('source', None)
-        # path = wsroot / name
-
-        # if not gitrepo.GitRepo.is_git_repo(path):
-        #    # TODO implement redownloading from remote
-        #    msg = "path '{}' is not a git repo even though it's in the manifest!".format(path)
-        #    raise Exception(msg)
-
-        return gitrepo.GitRepo(source, commit, name=name, wsroot=wsroot)
-
-    @staticmethod
-    def from_cwd(ws):
-        cwd = Path(os.getcwd()).resolve()
-
-        # Get the top-directory name relative to the workspace
-        name = cwd.relative_to(ws.path).parts[0]
-        package = ws.lock.get_package(name)
-        if package is None:
-            msg = "'{}' is not a package in workspace at '{}'".format(name, ws.path)
-            raise NotAPackageError(msg)
-        return package
-
-
-if __name__ == '__main__':
-    import doctest
-    doctest.testmod()
+
+        non_hash = len(self.unresolved_revision) < 40
+
+        if non_hash and not download:
+            log.error("Cannot create a reproducible workspace!")
+            raise WitBug("Cannot resolve '{}' without permission to download".format(non_hash))
```

Comment:
Shouldn't we check if we've already checked out the repo before erroring here? What if someone asks for a non_hash for something already in the workspace?

Reply:
Upstream may have different tags than local.

1. Bob downloads package `foo@origin/master`
2. Foo's remote moves `origin/master` to a different commit
3. Bob runs `wit update-pkg foo::origin/master`
4. Bob gets the manifest of the old `origin/master`
5. Alice, who has not downloaded `foo` before, runs `wit add-pkg foo:origin/master`
6. Bob and Alice have different workspaces
7. The workspace is non-reproducible (!)

If we don't fetch when our unresolved_revision is a tag, and the remote moves a tag, a user with the repo already cloned would get a different workspace than a user who has not yet downloaded the package.

Reply:
We may have downloaded an old version, but we need to have the latest version downloaded